### PR TITLE
Metadata text search api

### LINKED ISSUE #1592: Text search on transactions metadata
Similar to the text search provided via `/assets?search=` for assets, we'd like to have the same functionality for a transactions meta data.

Internally for BigchainDB this would mean:

- splitting meta data and transaction apart in the underlying data base
- content linking between the two
- implement an endpoint similar to `/asset?search=` for meta data

This use case was put forth by a high priority customer.

### PR DESCRIPTION
**Pull request by kansi**
Fixes #1592

### CODE REVIEW

File: bigchaindb/backend/mongodb/query.py [hunk 1]
```diff
@@ -348,9 +366,12 @@ def get_new_blocks_feed(conn, start_block_id):
 
 @register_query(MongoDBConnection)
 def text_search(conn, search, *, language='english', case_sensitive=False,
-                diacritic_sensitive=False, text_score=False, limit=0):
+                diacritic_sensitive=False, text_score=False, limit=0, table=None):
+    if table is None:
```
Comment (r-marques) [inline bigchaindb/backend/mongodb/query.py]: I think it would be cleaner and less error prone to make table a required parameter or at least change the default value of table from `None` to `assets` to eliminate the need for this `if` statement.

File: bigchaindb/core.py [hunk 2]
```diff
@@ -390,6 +395,13 @@ def get_spent(self, txid, output):
                     ' with the chain'.format(txid))
             # if its not and invalid transaction
             if status is not None:
+                if 'metadata' not in transaction:
```
Comment (r-marques) [inline bigchaindb/core.py]: Isn't this already handled by `Transaction.from_db` in `Bigchain.get_transaction`?

Comment (kansi) [reply bigchaindb/core.py]: @r-marques `transactions` here are fetched using `backend.query.get_spent`. So, they don't have metadata preloaded. We could assign transaction to a variable here `_, status = self.get_transaction ...` rather than ignoring which would help us avoid the `if`?

File: bigchaindb/web/views/metadata.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+from bigchaindb.backend.exceptions import OperationError
+from bigchaindb.web.views.base import make_error
+
+logger = logging.getLogger(__name__)
+
+
+class MetadataApi(Resource):
+    def get(self):
+        """API endpoint to perform a text search on transaction metadata.
```
Comment (ttmc) [inline bigchaindb/web/views/metadata.py]: This new HTTP API endpoint must be documented in the page about the HTTP API, similar to the documentation about the asset text search endpoint: https://docs.bigchaindb.com/projects/server/en/master/http-client-server-api.html#assets

File: bigchaindb/models.py [hunk 4]
```diff
@@ -352,11 +361,15 @@ def from_db(cls, bigchain, block_dict, from_dict_kwargs=None):
...
+        txn_ids = cls.get_txn_ids(block_dict)
+        metadata = bigchain.get_metadata(txn_ids)
+        # reconstruct block
         block_dict = cls.couple_assets(block_dict, assets)
+        block_dict = cls.couple_metadata(block_dict, metadata)
         kwargs = from_dict_kwargs or {}
         return cls.from_dict(block_dict, **kwargs)
 
-    def decouple_assets(self):
+    def decouple_assets(self, block_dict=None):
```
Comment (r-marques) [inline bigchaindb/models.py]: Why do we need that additional optional parameter?

File: bigchaindb/models.py [hunk 5]
```diff
@@ -365,6 +378,9 @@ def decouple_assets(self):
             the block being the dict of the block with no assets in the CREATE
             transactions.
         """
+        if block_dict is None:
```
Comment (r-marques) [inline bigchaindb/models.py]: This seems irrelevant. `block_dict = deepcopy(self.to_dict())` is always called regardless.

File: bigchaindb/models.py [hunk 6]
```diff
@@ -376,6 +392,28 @@ def decouple_assets(self):
...
+            tuple: (metadatas, block) with the metadatas being a list of dict/null and
+            the block being the dict of the block with no metadata in any transaction.
+        """
+        if block_dict is None:
+            block_dict = deepcopy(self.to_dict())
+
+        metadatas = []
+        for transaction in block_dict['block']['transactions']:
+            metadata = transaction.pop('metadata')
+            if isinstance(metadata, dict):
```
Comment (r-marques) [inline bigchaindb/models.py]: `metadata` is either `None` or a `dict`. This is enforced by the schema validation. So a check to see if it's not None should be enough, right?

File: bigchaindb/models.py [hunk 7]
```diff
@@ -376,6 +392,28 @@ def decouple_assets(self):
...
+                metadata.update({'id': transaction['id']})
```
Comment (r-marques) [inline bigchaindb/models.py]: We should probably nest the metadata under a new key or else if the metadata contains a key named `id` this will override it. We can probably have a data model similar to the asset. E.g.:
```python
{
     'id': '<transaction id>'
    'metadata': '<the metadata dict>'
}
```

File: bigchaindb/models.py [hunk 8]
```diff
@@ -402,6 +440,39 @@ def couple_assets(block_dict, assets):
...
+                a database call.
+
+        Returns:
+            dict: The dict of the reconstructed block.
+        """
+        # create a dict with {'<txid>': metadata}
+        metadatal = {m.pop('id'): m for m in metadatal}
+        # add the metadata to their corresponding transactions
+        for transaction in block_dict['block']['transactions']:
+            if 'metadata' not in transaction:
```
Comment (r-marques) [inline bigchaindb/models.py]: `metadata` should not be in the transaction at this point.

File: bigchaindb/models.py (hunk 7 above)
Comment (kansi) [reply bigchaindb/models.py]: agreed.

File: bigchaindb/models.py (hunk 4 above)
Comment (kansi) [reply bigchaindb/models.py]: this is done in order to make the interface same as `decouple_metadata`.

File: bigchaindb/core.py (hunk 2 above)
Comment (r-marques) [reply bigchaindb/core.py]: I think it would be better to use the transaction returned by `get_transaction`. This way we reduce the number of database calls and we eliminate the need to handle the metadata.

File: bigchaindb/models.py (hunk 5 above)
Comment (kansi) [reply bigchaindb/models.py]: That is true as we have only one place where `decouple_assets` is called but I tired to generalize and make the interface for `decouple_assets` and `decouple_metadata` similar. We can do it either way

File: bigchaindb/web/views/metadata.py (hunk 3 above)
Comment (kansi) [reply bigchaindb/web/views/metadata.py]: Sure.

File: bigchaindb/models.py (hunk 5 above)
Comment (r-marques) [reply bigchaindb/models.py]: That is ok but the logic here is still wrong. `block_dict = deepcopy(self.to_dict())` is always called  and the `if` statement is irrelevant.

Comment (kansi) [reply bigchaindb/models.py]: @r-marques Its my bad, I didn't realize you were talking about [this](https://github.com/bigchaindb/bigchaindb/pull/1812/files#diff-9f1ab5b5daf1b3072e2123b6939bfb07R383)?

**Comment by ttmc** [comment]
@kansi I don't know why codecov is upset about a supposed 0.06% change in coverage. I wouldn't worry about it.

**Comment by kansi** [comment]
@ttmc sure, just need a final review from @r-marques then I will merge it.
